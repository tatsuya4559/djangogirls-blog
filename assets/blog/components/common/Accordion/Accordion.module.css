.check {
  display: none;
}

.label {
  display: flex;
}

.iconWrapper {
  flex: 1 1 auto;
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.icon {
  transition: all 0.2s;
}

.item {
  max-height: 0;
  opacity: 0;
  transition: all 0.3s;
  overflow: hidden;
}

.check:checked + .label + .item {
  max-height: 100vh;
  opacity: 1;
}

.check:checked + .label .icon {
  transform: rotate(-180deg);
}
