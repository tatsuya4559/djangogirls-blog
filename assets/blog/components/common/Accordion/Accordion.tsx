import React, { useState, useContext, useMemo } from 'react';
import { FaChevronDown } from 'react-icons/fa';
import uniqueId from 'lodash/uniqueId';
import cx from 'classnames';
import classes from './Accordion.module.css';

const AccordionContext = React.createContext('accordionNamespace');

type HeaderProps = {
  defaultOpen?: boolean;
  className?: string;
};

const AccordionHeader: React.FC<HeaderProps> = ({
  defaultOpen,
  className,
  children,
}) => {
  const name = useContext(AccordionContext);
  const id = useMemo(() => uniqueId('accordion'), []);
  return (
    <>
      <input
        className={classes.check}
        type="radio"
        name={name}
        id={id}
        defaultChecked={defaultOpen}
      />
      <label className={cx(classes.label, className)} htmlFor={id}>
        {children}
        <div className={classes.iconWrapper}>
          <FaChevronDown className={classes.icon} />
        </div>
      </label>
    </>
  );
};

type ItemProps = {
  className?: string;
};

const AccordionItem: React.FC<ItemProps> = ({ className, children }) => (
  <div className={cx(classes.item, className)}>{children}</div>
);

interface IAccordion extends React.FC {
  Header: typeof AccordionHeader;
  Item: typeof AccordionItem;
}

const Accordion: IAccordion = ({ children }) => {
  const name = useMemo(() => uniqueId('accordionNamespace'), []);
  return (
    <AccordionContext.Provider value={name}>
      {children}
    </AccordionContext.Provider>
  );
};

Accordion.Header = AccordionHeader;
Accordion.Item = AccordionItem;

export default Accordion;
