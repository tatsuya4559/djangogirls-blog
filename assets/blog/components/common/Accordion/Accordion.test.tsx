import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Accordion from './Accordion';

afterEach(cleanup);

test('snapshot', () => {
  const { asFragment } = render(
    <Accordion>
      <Accordion.Header>hoge</Accordion.Header>
      <Accordion.Content>hoge content</Accordion.Content>
      <Accordion.Header>fuga</Accordion.Header>
      <Accordion.Content>fuga content</Accordion.Content>
    </Accordion>,
  );
  expect(asFragment()).toMatchSnapshot();
});

describe('Accordion.Content', () => {
  test('childrenが描画されること', () => {
    const childText = 'childです';
    const { getByText } = render(
      <Accordion.Content>
        <p>{childText}</p>
      </Accordion.Content>,
    );
    expect(getByText(childText)).toBeInTheDocument();
  });

  test('classNameが適用されること', () => {
    const { getByText } = render(
      <Accordion.Content className="hoge">fuga</Accordion.Content>,
    );
    expect(getByText('fuga')).toHaveClass('hoge');
  });
});

describe('Accordion.Header', () => {
  test('childrenが描画されること', () => {
    const childText = 'hoge';
    const { getByText } = render(
      <Accordion.Header>
        <p>{childText}</p>
      </Accordion.Header>,
    );
    expect(getByText(childText)).toBeInTheDocument();
  });

  test('classNameが適用されること', () => {
    const { getByText } = render(
      <Accordion.Header className="hoge">fuga</Accordion.Header>,
    );
    expect(getByText('fuga')).toHaveClass('hoge');
  });

  test('Headerごとにラジオボタンのidが異なること', () => {
    const { getByLabelText } = render(
      <>
        <Accordion.Header>hoge</Accordion.Header>
        <Accordion.Header>fuga</Accordion.Header>
      </>,
    );
    const hogeId = getByLabelText('hoge').getAttribute('id');
    const fugaId = getByLabelText('fuga').getAttribute('id');
    expect(hogeId).not.toBe(fugaId);
  });

  test('defaultOpenが指定されたHeaderはcheckがついていること', () => {
    const { getByLabelText } = render(
      <>
        <Accordion.Header defaultOpen>hoge</Accordion.Header>
        <Accordion.Header>fuga</Accordion.Header>
      </>,
    );
    expect(getByLabelText('hoge')).toBeChecked();
    expect(getByLabelText('fuga')).not.toBeChecked();
  });

  test('labelをクリックしてcheckがつくこと', () => {
    const { getByText, getByLabelText } = render(
      <>
        <Accordion.Header>hoge</Accordion.Header>
        <Accordion.Header>fuga</Accordion.Header>
      </>,
    );
    expect(getByLabelText('hoge')).not.toBeChecked();
    expect(getByLabelText('fuga')).not.toBeChecked();

    fireEvent.click(getByText('hoge'));

    expect(getByLabelText('hoge')).toBeChecked();
    expect(getByLabelText('fuga')).not.toBeChecked();

    fireEvent.click(getByText('fuga'));

    expect(getByLabelText('hoge')).not.toBeChecked();
    expect(getByLabelText('fuga')).toBeChecked();
  });
});

describe('Accordion', () => {
  test('Accordion内でラジオボタンのグループが揃っていること', () => {
    const { getByLabelText } = render(
      <>
        <Accordion>
          <Accordion.Header>hoge</Accordion.Header>
          <Accordion.Header>fuga</Accordion.Header>
        </Accordion>
        <Accordion>
          <Accordion.Header>piyo</Accordion.Header>
        </Accordion>
      </>,
    );
    const hogeName = getByLabelText('hoge').getAttribute('name');
    const fugaName = getByLabelText('fuga').getAttribute('name');
    const piyoName = getByLabelText('piyo').getAttribute('name');
    expect(hogeName).toBe(fugaName);
    expect(hogeName).not.toBe(piyoName);
  });
});
