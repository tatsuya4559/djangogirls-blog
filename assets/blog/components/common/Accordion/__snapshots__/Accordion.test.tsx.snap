// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<DocumentFragment>
  <input
    class="check"
    id="accordion2"
    name="accordionNamespace1"
    type="radio"
  />
  <label
    class="label"
    for="accordion2"
  >
    hoge
    <div
      class="iconWrapper"
    >
      <svg
        class="icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </div>
  </label>
  <div
    class="content"
  >
    hoge content
  </div>
  <input
    class="check"
    id="accordion3"
    name="accordionNamespace1"
    type="radio"
  />
  <label
    class="label"
    for="accordion3"
  >
    fuga
    <div
      class="iconWrapper"
    >
      <svg
        class="icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="6 9 12 15 18 9"
        />
      </svg>
    </div>
  </label>
  <div
    class="content"
  >
    fuga content
  </div>
</DocumentFragment>
`;
